src/Module.ts

```diff
@@ -148,7 +134,7 @@ export default class Module {
 	chunk: Chunk;
 
 	ast: Program;
-	private astClone: Program;
+	private astClone: ESTree.Program;
```

Comment:
This confuses me. Shouldn't `ast` and `astClone` have the same type?

Reply:
There are really three things here not two - `ast`, `astClone` and `enhancedAst`. `ast` doubles as both EStree and the enhanced form, while the cache API for ast and astClone relies only on the EStree interface.

Reply:
Interesting. Should `ast` be defined as `Program & ESTree.Program` then?

Reply:
Program is a superset of ESTree.Program, so that would be the same thing here I think.

Reply:
I think that is the confusing part to me, because the `ast` that is passed to `setSource` is defined as `ESTree.Program` (via `ModuleJSON`). So assigning it to a field of type `Program`, which is supposed to be a superset, is technically an error then which is hidden by the [type cast](https://github.com/rollup/rollup/pull/2108/files/794bd7464a7bb90c84d4d18b47f0296247561105#diff-356410863c3a18aee7d09896d52d6b08R199), right?

Reply:
The way to think about this is that `Module.ast` when first assigned is an `ESTree.Program` but transitions to a `Rollup.Program` during the enhance call (initiated by analyse). When `Module.ast` is exported, cloned or copied, even though it is a `Rollup.Program` it is only needed to be a `ESTree.Program` through the cache interfaces, as later on enhance will be called again on it.

That's why I say it's really two things - `Module.enhancedAst` and `Module.ast`.

Reply:
Note this problem is only temporary as I am currently working on a branch that will finally separate these two types. In the end, `Module.ast` will be only a `Program` while the `Program` constructor will receive an `ESTree.Program` as parameter.

Reply:
Hm, that's tricky. In my opinion, it should really be `ESTree.Program | Program` then and consequently all five compiler errors that result from this change should add a guard so that this won't lead to a runtime error in the future when someone changes how and when the different methods in `Module` are called for some reason, e.g.:

```ts
if ('bind' in node) {
	node.bind();
}
```

or:

```ts
if (!('bind' in node)) {
	throw new Error('Unexpected AST node found. ...');
}
```